core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/streams/streams.go

```diff
@@ -177,65 +176,97 @@ func (s *streams) buildResult(ctx context.Context, i int, checkResult ocr2keeper
-		checkResults[i].Retryable = retryable
-		checkResults[i].RetryInterval = retryInterval
-		checkResults[i].PipelineExecutionState = uint8(state)
-		checkResults[i].IneligibilityReason = uint8(reason)
-		return
+		s.lggr.Errorf("at block %d upkeep %s checkCallback packing err: %s", lookup.Block, lookup.UpkeepId, err.Error())
+		checkResults[i].Retryable = false
+		checkResults[i].PipelineExecutionState = uint8(mercury.PackUnpackDecodeFailed)
+		return err
 	}
 
-	for j, v := range values {
-		s.lggr.Infof("at block %d upkeep %s requested time %s doMercuryRequest values[%d]: %s", lookup.Block, lookup.UpkeepId, lookup.Time, j, hexutil.Encode(v))
+	var mercuryBytes hexutil.Bytes
+	args := map[string]interface{}{
+		"to":   s.registry.Address().Hex(),
+		"data": hexutil.Bytes(payload),
 	}
 
-	state, retryable, mercuryBytes, err := s.checkCallback(ctx, values, lookup)
-	if err != nil {
+	// call checkCallback function at the block which OCR3 has agreed upon
+	if err = s.client.CallContext(ctx, &mercuryBytes, "eth_call", args, hexutil.EncodeUint64(lookup.Block)); err != nil {
 		s.lggr.Errorf("at block %d upkeep %s checkCallback err: %s", lookup.Block, lookup.UpkeepId, err.Error())
-		checkResults[i].Retryable = retryable
-		checkResults[i].PipelineExecutionState = uint8(state)
-		return
+		checkResults[i].Retryable = true
+		checkResults[i].PipelineExecutionState = uint8(mercury.RpcFlakyFailure)
+		return err
 	}
+
 	s.lggr.Infof("at block %d upkeep %s requested time %s checkCallback mercuryBytes: %s", lookup.Block, lookup.UpkeepId, lookup.Time, hexutil.Encode(mercuryBytes))
 
 	unpackCallBackState, needed, performData, failureReason, _, err := s.packer.UnpackCheckCallbackResult(mercuryBytes)
 	if err != nil {
 		s.lggr.Errorf("at block %d upkeep %s requested time %s UnpackCheckCallbackResult err: %s", lookup.Block, lookup.UpkeepId, lookup.Time, err.Error())
 		checkResults[i].PipelineExecutionState = unpackCallBackState
-		return
+		return err
 	}
 
 	if failureReason == uint8(mercury.MercuryUpkeepFailureReasonMercuryCallbackReverted) {
 		checkResults[i].IneligibilityReason = uint8(mercury.MercuryUpkeepFailureReasonMercuryCallbackReverted)
 		s.lggr.Debugf("at block %d upkeep %s requested time %s mercury callback reverts", lookup.Block, lookup.UpkeepId, lookup.Time)
-		return
+		return fmt.Errorf("at block %d upkeep %s requested time %s mercury callback reverts", lookup.Block, lookup.UpkeepId, lookup.Time)
+
 	}
 
 	if !needed {
 		checkResults[i].IneligibilityReason = uint8(mercury.MercuryUpkeepFailureReasonUpkeepNotNeeded)
 		s.lggr.Debugf("at block %d upkeep %s requested time %s callback reports upkeep not needed", lookup.Block, lookup.UpkeepId, lookup.Time)
-		return
+		return fmt.Errorf("at block %d upkeep %s requested time %s callback reports upkeep not needed", lookup.Block, lookup.UpkeepId, lookup.Time)
```

Comment:
we'll double log this as of now.

debug above
and error in the new function you introduced

we shouldn't add error logs for this purpose as that should indicate something wrong in the node

Reply:
Indeed, we are double logging. 

How about we remove the `s.lggr.Errorf()` (e.g. L192, L206, L216) in this new `CheckCallback` function, and let the caller capture the error and print (in doLookup() L #184)?

Reply:
that should work, will just need to verify all logs in existing cases and ensure they remain the same